dm/syncer/ddl.go

```diff
@@ -18,18 +18,935 @@ import (
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
 )
 
+type ShardDDLStrategy interface {
+	preFilter(ddlInfo *ddlInfo, qec *queryEventContext, sourceTable *filter.Table, targetTable *filter.Table) (bool, error)
+	handleDDL(qec *queryEventContext) error
+}
+
+type ShardDDL struct {
+	logger log.Logger
+
+	strategy ShardDDLStrategy
+
+	binlogFilter               *bf.BinlogEvent
+	metricsProxies             *metrics.Proxies
+	name                       string
+	workerName                 string
+	sourceID                   string
+	enableGTID                 bool
+	shardMode                  string
+	upstreamTZStr              string
+	onlineDDL                  onlineddl.OnlinePlugin
+	checkpoint                 CheckPoint
+	tableRouter                *regexprrouter.RouteTable
+	sourceTableNamesFlavor     utils.LowerCaseTableNamesFlavor
+	collationCompatible        string
+	charsetAndDefaultCollation map[string]string
+	idAndCollationMap          map[int]string
+	baList                     *tablefilter.Filter
+
+	recordSkipSQLsLocation func(ec *eventContext) error
+	trackDDL               func(usedSchema string, trackInfo *ddlInfo, ec *eventContext) error
+	saveTablePoint         func(table *filter.Table, location binlog.Location)
+	flushJobs              func() error
+}
+
+// NewShardDDL creates a new ShardDDL instance.
+func NewShardDDL(pLogger *log.Logger, syncer *Syncer) *ShardDDL {
+	shardDDL := &ShardDDL{
+		logger:                     pLogger.WithFields(zap.String("component", "ddl")),
+		binlogFilter:               syncer.binlogFilter,
+		metricsProxies:             syncer.metricsProxies,
+		name:                       syncer.cfg.Name,
+		workerName:                 syncer.cfg.WorkerName,
+		sourceID:                   syncer.cfg.SourceID,
+		enableGTID:                 syncer.cfg.EnableGTID,
+		shardMode:                  syncer.cfg.ShardMode,
+		upstreamTZStr:              syncer.upstreamTZStr,
+		onlineDDL:                  syncer.onlineDDL,
+		checkpoint:                 syncer.checkpoint,
+		tableRouter:                syncer.tableRouter,
+		sourceTableNamesFlavor:     syncer.SourceTableNamesFlavor,
+		collationCompatible:        syncer.cfg.CollationCompatible,
+		charsetAndDefaultCollation: syncer.charsetAndDefaultCollation,
+		idAndCollationMap:          syncer.idAndCollationMap,
+		baList:                     syncer.baList,
+		recordSkipSQLsLocation:     syncer.recordSkipSQLsLocation,
+		trackDDL:                   syncer.trackDDL,
+		saveTablePoint:             syncer.saveTablePoint,
+		flushJobs:                  syncer.flushJobs,
+	}
+	switch syncer.cfg.ShardMode {
+	case config.ShardPessimistic:
+		shardDDL.strategy = NewPessimistDDL(&shardDDL.logger, syncer)
+	case config.ShardOptimistic:
+		shardDDL.strategy = NewOptimistDDL(&shardDDL.logger, syncer)
+	default:
+		shardDDL.strategy = NewNormalDDL(&shardDDL.logger, syncer)
+	}
+	return shardDDL
+}
+
+type Normal struct {
+	ShardDDLStrategy
```

Comment:
why embedded this interface? we can assert Normal/Optimist/Pessimist implement `ShardDDLStrategy`